### PR TITLE
PP-4373 Simplify auth 3ds interface

### PR DESCRIPTION
**Pull request by mrlumbu**
This PR does two things:
- Pushes some functionality in `Auth3DSResponseService` into `CarAuthoriseBaseService` - basically logging and metrics, which we can also use from any other authorise service
- Simplifies the interface of the payment provider method `authorise3dsResponse` to respond with a simpler object, that only contains the necessary information for doing the things that the service and resource need to do. This is still not perfect, but is massively simpler and cleaner, and can be refined in further work.

### CODE REVIEW

File: src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProvider3dsTest.java [hunk 1]
```diff
@@ -34,20 +36,20 @@ public void shouldRequire3dsAuthoriseRequest() {
     @Test
     public void shouldAuthorise3dsResponseIfMatchesWithEpdqStatus() {
         mockPaymentProviderResponse(200, successAuthorisedQueryResponse());
-        GatewayResponse<BaseAuthoriseResponse> response = provider.authorise3dsResponse(buildTestAuthorisation3dsVerifyRequest("AUTHORISED"));
+        Gateway3DSAuthorisationResponse response = provider.authorise3dsResponse(buildTestAuthorisation3dsVerifyRequest("AUTHORISED"));
         verifyPaymentProviderRequest(successAuthQueryRequest());
         assertTrue(response.isSuccessful());
-        assertThat(response.getBaseResponse().get().authoriseStatus(), is(AUTHORISED));
+        assertThat(response.isSuccessful(), is(true));
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProvider3dsTest.java]: I think this line can be removed as is the same as above.

File: src/test/java/uk/gov/pay/connector/paymentprocessor/service/Card3dsResponseAuthServiceTest.java [hunk 2]
```diff
@@ -80,14 +77,7 @@ public void setupMockExecutorServiceMock() {
...
     private void setupPaymentProviderMock(String transactionId, AuthoriseStatus authoriseStatus, String errorCode, ArgumentCaptor<Auth3dsResponseGatewayRequest> argumentCaptor) {
-        WorldpayOrderStatusResponse worldpayResponse = mock(WorldpayOrderStatusResponse.class);
-        when(worldpayResponse.getTransactionId()).thenReturn(transactionId);
-        when(worldpayResponse.authoriseStatus()).thenReturn(authoriseStatus);
-        when(worldpayResponse.getErrorCode()).thenReturn(errorCode);
-        GatewayResponseBuilder<WorldpayOrderStatusResponse> gatewayResponseBuilder = responseBuilder();
-        GatewayResponse authorisationResponse = gatewayResponseBuilder
-                .withResponse(worldpayResponse)
-                .build();
+        Gateway3DSAuthorisationResponse authorisationResponse = new Gateway3DSAuthorisationResponse(authoriseStatus, null);
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/connector/paymentprocessor/service/Card3dsResponseAuthServiceTest.java]: `errorCode` is never used. `transactionId` is never used but should replace `null` in the call to constructor

File: src/main/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProvider.java [hunk 3]
```diff
@@ -149,11 +150,16 @@ public PaymentGatewayName getPaymentGatewayName() {
...
                     authoriseStatus.name()))
                     .inc();
-            return reconstructErrorBiasedGatewayResponse(gatewayResponse, authoriseStatus, auth3DResult);
+            gatewayResponse = reconstructErrorBiasedGatewayResponse(gatewayResponse, authoriseStatus, auth3DResult);
 
         }
-        return gatewayResponse;
-    }
+        
+        if (gatewayResponse.getBaseResponse().isPresent()) {
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProvider.java]: This whole block can be replaced with functional style Java:
```
return gatewayResponse.getBaseResponse().map(baseResponse -> new Gateway3DSAuthorisationResponse(baseResponse.authoriseStatus(), baseResponse.getTransactionId()))
                .orElseGet(Gateway3DSAuthorisationResponse::ofException);
```

File: src/main/java/uk/gov/pay/connector/gateway/model/response/Gateway3DSAuthorisationResponse.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public ChargeStatus getMappedChargeStatus() {
+        return authorisationStatus.getMappedChargeStatus();
+    }
+
+    public Optional<String> getTransactionId() {
+        return Optional.ofNullable(transactionId);
+    }
+}
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/connector/gateway/model/response/Gateway3DSAuthorisationResponse.java:50]: A factory method that will build objects for exceptions would be handy. I see in many places calling the constructor with exception and null. Something like bellow will get rid of the call with null and having to pass the same exception status. 
```
public static Gateway3DSAuthorisationResponse ofException() {
        return new Gateway3DSAuthorisationResponse(BaseAuthoriseResponse.AuthoriseStatus.EXCEPTION, null);
    }
```

File: src/main/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProvider.java [hunk 5]
```diff
@@ -64,11 +65,9 @@ public SandboxPaymentProvider() {
     }
 
     @Override
-    public GatewayResponse<BaseAuthoriseResponse> authorise3dsResponse(Auth3dsResponseGatewayRequest request) {
-        GatewayResponseBuilder<BaseResponse> gatewayResponseBuilder = responseBuilder();
-        return gatewayResponseBuilder
-                .withGatewayError(new GatewayError("3D Secure not implemented for Sandbox", GENERIC_GATEWAY_ERROR))
-                .build();
+    public Gateway3DSAuthorisationResponse authorise3dsResponse(Auth3dsResponseGatewayRequest request) {
+        return new Gateway3DSAuthorisationResponse(BaseAuthoriseResponse.AuthoriseStatus.EXCEPTION, null);
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProvider.java]: Calling the constructor with null can be replaced with wither constructor chaining or a factory method as suggested in this class.

File: src/main/java/uk/gov/pay/connector/gateway/smartpay/SmartpayPaymentProvider.java [hunk 6]
```diff
@@ -69,9 +70,15 @@ public PaymentGatewayName getPaymentGatewayName() {
     }
 
     @Override
-    public GatewayResponse<BaseAuthoriseResponse> authorise3dsResponse(Auth3dsResponseGatewayRequest request) {
+    public Gateway3DSAuthorisationResponse authorise3dsResponse(Auth3dsResponseGatewayRequest request) {
         Either<GatewayError, GatewayClient.Response> response = client.postRequestFor(null, request.getGatewayAccount(), build3dsResponseAuthOrderFor(request));
-        return GatewayResponseGenerator.getSmartpayGatewayResponse(client, response, SmartpayAuthorisationResponse.class);
+        GatewayResponse<BaseAuthoriseResponse> gatewayResponse = GatewayResponseGenerator.getSmartpayGatewayResponse(client, response, SmartpayAuthorisationResponse.class);
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/connector/gateway/smartpay/SmartpayPaymentProvider.java:75]: This whole block can be replaced with functional style Java, as per example given in Epdq.

File: src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProvider.java [hunk 7]
```diff
@@ -87,9 +88,15 @@ public PaymentGatewayName getPaymentGatewayName() {
     }
 
     @Override
-    public GatewayResponse<BaseAuthoriseResponse> authorise3dsResponse(Auth3dsResponseGatewayRequest request) {
+    public Gateway3DSAuthorisationResponse authorise3dsResponse(Auth3dsResponseGatewayRequest request) {
         Either<GatewayError, GatewayClient.Response> response = authoriseClient.postRequestFor(null, request.getGatewayAccount(), build3dsResponseAuthOrder(request));
-        return GatewayResponseGenerator.getWorldpayGatewayResponse(authoriseClient, response, WorldpayOrderStatusResponse.class);
+        GatewayResponse<BaseAuthoriseResponse> gatewayResponse = GatewayResponseGenerator.getWorldpayGatewayResponse(authoriseClient, response, WorldpayOrderStatusResponse.class);
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProvider.java:93]: This whole block can be replaced with functional style Java, as per example given in Epdq.

**Review by georgeracu — CHANGES_REQUESTED** [review]
Only a few changes to simplify the code. Other than that 👍 😸

File: src/main/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProvider.java (hunk 3 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProvider.java]: 👍

File: src/main/java/uk/gov/pay/connector/gateway/model/response/Gateway3DSAuthorisationResponse.java (hunk 4 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/model/response/Gateway3DSAuthorisationResponse.java:50]: 👍Constructor is private now and using factory methods at all places.

File: src/main/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProvider.java (hunk 5 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProvider.java]: 👍 Added factory method

File: src/main/java/uk/gov/pay/connector/gateway/smartpay/SmartpayPaymentProvider.java (hunk 6 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/smartpay/SmartpayPaymentProvider.java:75]: 👍 Updated

File: src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProvider.java (hunk 7 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProvider.java:93]: 👍 Updated

File: src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProvider3dsTest.java (hunk 1 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProvider3dsTest.java]: Good spot :). Removed.

File: src/test/java/uk/gov/pay/connector/paymentprocessor/service/Card3dsResponseAuthServiceTest.java (hunk 2 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/paymentprocessor/service/Card3dsResponseAuthServiceTest.java]: Removed `errorCode` and replaced null with `transactionId`

**Review by kbottla — COMMENTED** [review]
👍 Thanks @georgeracu. Addressed all comments

**Review by georgeracu — APPROVED** [review]
🎉

**Pull request merged**